### PR TITLE
TECH : use the stakhanov package and remove all duplicated code

### PR DESCRIPTION
**Pull request by Tehem**
### Purpose of this PR

- Use stakhanov package
- Remove all duplicated code

### Checks

- [ ] Documentation of environment variables is very clear, even for a newcomer
- [ ] All relevant usecases of are tested
- [ ] These changes won't cause memory / CPU problems even if business grows by a factor of 100
- [ ] It's ok if these changes go to production and then are rolled-back

### CODE REVIEW

File: lib/createWorker.js [hunk 1]
```diff
@@ -18,7 +13,11 @@ const {
  */
 function formatCreateWorkersParameters(handler, config, options) {
   return {
-    handlers: [{ handle: handler, validate: options.validator, routingKey: config.routingKey }],
+    handlers: [{
+      handle: handler,
+      validate: options.validator || _.identity,
```
Comment (lledey) [inline lib/createWorker.js]: Please remove the default value. Schema validation must absolutely be required for AMQP handlers, like it is for HTTP controllers

File: lib/createWorker.js [hunk 2]
```diff
@@ -28,32 +27,18 @@ function formatCreateWorkersParameters(handler, config, options) {
...
- * @param {function} [options.validator] function to validate the message with
- * @param {Number} [options.heartbeat] to override default heartbeat
- * @param {Number} [options.taskTimeout] to override default task timeout
- * @param {Number} [options.processExitTimeout] to override default process exit timeout
- * @param {Number} [options.channelPrefetch] to override default channel prefetch value
  * @returns {Object} a worker instance with connection, channel, and listen/close functions
- * @deprecated
+ * @deprecated use createWorkers instead
  */
 function createWorker(handler, config, options = {}) {
```
Comment (lledey) [inline lib/createWorker.js]: What do you think if we completely remove this legacy function and bump a major version?

File: lib/createWorker.js [hunk 3]
```diff
@@ -28,32 +27,18 @@ function formatCreateWorkersParameters(handler, config, options) {
  * Create a worker instance based on configuration
  * @param {function} handler function to handle each message
  * @param {Object} config configuration for the worker
- * @param {String} config.workerName name for the worker
```
Comment (lledey) [inline lib/createWorker.js]: Why did you remove that?

File: lib/createWorkers.js [hunk 4]
```diff
@@ -1,350 +1,19 @@
...
-  applyConfiguration,
-  applyOptions
-} = require('./schema/createWorkers.schema');
-
-const DEFAULT_EXCHANGE_TYPE = 'topic';
-
 /**
  * Create a worker instance based on configuration
  * @param {Array} handlers array of handlers to handle each message
- * @param {function} handlers.handle function to handle each message
```
Comment (lledey) [inline lib/createWorkers.js]: Why did you remove this doc?

File: package.json [hunk 5]
```diff
@@ -10,18 +10,14 @@
...
   },
   "dependencies": {
-    "amqplib": "0.5.1",
     "chpr-logger": "2.4.1",
-    "chpr-metrics": "1.1.0",
-    "co": "4.6.0",
-    "joi": "10.2.1",
-    "lodash": "4.14.2"
+    "lodash": "4.14.2",
+    "stakhanov": "0.2.1"
```
Comment (lledey) [inline package.json]: It doesn't seem very safe to introduce a dependency with a beta version (even if it's our code)

File: test/lib/worker.test.js [hunk 6]
```diff
@@ -482,8 +482,8 @@ describe('Worker library', () => {
         processExitTimeout: 1
       });
       yield worker.listen();
-      yield worker.close();
-      yield worker.wait('worker.closed');
+      yield worker.close(); // true by default
+      yield cb => setTimeout(cb, 1000);
```
Comment (lledey) [inline test/lib/worker.test.js]: The test was better before, 1 second is very long for a unit test, and the test can now fail randomly on slow machines

Comment (Tehem) [reply test/lib/worker.test.js]: Yes but it doesn't work anymore... the event is triggered before the wait could catch it. I'm just following the same approach as the one in stakhanov : 
https://github.com/ChauffeurPrive/stakhanov/blob/master/test/lib/worker.test.js#L412-L414

**Comment by Tehem** [comment]
And for the record, the OS channel is a channel to which everyone was invited to join when we created it. If you weren't in it, it's just you did not answer the call :)

File: lib/createWorkers.js [hunk 7]
```diff
@@ -1,350 +1,19 @@
...
- *
- * It contains the base queue name and the routing key.
- *
- * @param       {String} queueName The base queue name.
- * @param       {String} routingKey The routing key.
- * @returns     {String} The formatted queue name.
- */
-function _getQueueName(queueName, routingKey) {
-  return `${queueName}.${routingKey}`;
+  options.logger = logger;
```
Comment (jean-k-cp) [inline lib/createWorkers.js]: you should not modify the `options` parameter, this side effect can be completely avoided 🙂

Comment (Tehem) [reply lib/createWorkers.js]: ok

File: lib/createWorkers.js (hunk 4 above)
Comment (lledey) [reply lib/createWorkers.js]: When implementing a worker I only want to look at the packages documentation I directly use, not the underlying packages. I would keep the doc here since we are going to update the code anyway if we bump a major version of our dependencies

File: package.json (hunk 5 above)
Comment (lledey) [reply package.json]: Bump to `0.3.0` after @jean-k-cp's fix in stakhanov

File: package.json [hunk 8]
```diff
@@ -10,18 +10,14 @@
...
   },
   "devDependencies": {
     "chai": "3.5.0",
     "co-mocha": "1.1.3",
     "co-supertest": "0.0.10",
```
Comment (lledey) [inline package.json]: Is it really used?

File: test/lib/createWorkers.test.js [hunk 9]
```diff
@@ -326,7 +271,7 @@ describe('Worker library', () => {
       expect(message).to.be.false();
     });
 
-    it('should call provided handlers and ack if handlers runs ok', function* test() {
+    it('should call all provided handlers and ack if handlers runs ok', function* test() {
```
Comment (lledey) [inline test/lib/createWorkers.test.js]: We are testing the underlying lib here. How about just calling `createWorkers` and checking that we passed the logger and the returned object?

File: package.json (hunk 5 above)
Comment (jean-k-cp) [reply package.json]: we'll do this in a next version so we have both versions: one with stakhanov (no graceful shutdown) and one with the extra graceful shutdown, and we can bump to the one we think is safe 🙂 🛩

File: lib/createWorkers.js [hunk 10]
```diff
@@ -12,8 +12,8 @@ const logger = require('chpr-logger');
  * @see https://github.com/ChauffeurPrive/stakhanov/blob/master/README.md for documentation
  */
 function createWorkers(handlers, config, options = {}) {
-  options.logger = logger;
-  return stakhanov.createWorkers(handlers, config, options);
+  const optionsWithLogger = Object.assign({}, options, { logger });
```
Comment (jean-k-cp) [inline lib/createWorkers.js]: 🍾 🎉

File: lib/createWorkers.js (hunk 4 above)
Comment (jean-k-cp) [reply lib/createWorkers.js]: I agree that duplication is a pain, but the jsdoc will help the IDE to suggest useful insights.
(note to myself: I'll have to document the new options when bumping this...)

File: package.json [hunk 11]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "chpr-worker",
-  "version": "1.2.0",
+  "version": "2.0.0",
```
Comment (jean-k-cp) [inline package.json]: please use 2.2 for the hotfix https://github.com/ChauffeurPrive/stakhanov/commits/0.2.2 :)

File: test/lib/createWorkers.test.js (hunk 9 above)
Comment (Tehem) [reply test/lib/createWorkers.test.js]: well, since the test are here I will leave them here, might be useful for non-reg...

File: package.json (hunk 11 above)
Comment (Tehem) [reply package.json]: done
